shoop/campaigns/modules.py

```diff
@@ -122,3 +130,16 @@ def get_discount_line(campaign, amount, price_so_far):
         if best_discount is not None:
             lines.append(get_discount_line(best_discount_campaign, best_discount, price_so_far))
         return lines
+
+    def _handle_line_effects(self, matching_campaigns, order_source, original_lines):
+        lines = []
+        for campaign in matching_campaigns:
+            effect = campaign.effects.first()
+            if not effect:
+                continue
+
+            if not hasattr(effect, "get_discount_lines"):
+                continue
+
+            lines += effect.get_discount_lines(order_source=order_source, original_lines=original_lines)
```

Comment:
Should use `lines.extend`, because the effect could return an iterable that is not a list, e.g. yield the lines.